src/rust/engine/fs/store/src/lib.rs

```diff
@@ -791,19 +791,26 @@ impl Store {
       .load_file_bytes_with(
         digest,
         move |bytes| {
-          OpenOptions::new()
-            .create(true)
-            .write(true)
-            .mode(if is_executable { 0o755 } else { 0o644 })
-            .open(&destination)
-            .and_then(|mut f| {
-              f.write_all(&bytes)?;
-              // See `materialize_directory`, but we fundamentally materialize files for other
-              // processes to read; as such, we must ensure data is flushed to disk and visible
-              // to them as opposed to just our process.
-              f.sync_all()
-            })
-            .map_err(|e| format!("Error writing file {:?}: {:?}", destination, e))
+          if destination.exists() {
+            std::fs::remove_file(&destination)
+          } else {
+            Ok(())
+          }
+          .and_then(|_| {
+            OpenOptions::new()
+              .create(true)
+              .write(true)
+              .mode(if is_executable { 0o755 } else { 0o644 })
+              .open(&destination)
```

Comment:
Does rust not have a way to "delete and then rewrite from 0"? Usually that's the default...

Reply:
https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.truncate

Reply:
I'm almost positive this is another half measure only making the conflict time window smaller, not 0. That is probably a useful advantage, except that I think this allows silent corruption for concurrent writes. The prior status quo failed fast. The change as-is fails fast again, but with a smaller chance. The rename attack would be concurrent safe with just last wins silently, 1st loses silently. This option is usually win but may corrupt silently... IIUC.

Reply:
Yup this works

Reply:
It may work, but, per my comment, I'm pretty sure it allows silent corruption. If you agree, I think we can't do this.

Reply:
I see benefits to each approach. I suppose that the number of syscalls is likely the same either way, and fast-failing is beneficial. I just hadn't thought about the benefit of failing fast here in case of a race, and was aiming for shorter code.

Reply:
I'm at a loss. One approach fails fast, the other can lead to silent corruption. Of those two don't you have to pick non corruption every time?

Reply:
Other components of this API are supposed to be defending against concurrent usage: it defends against overlapping directories, for example.

Having said that, those can all fail. So yea, fine. Fail fast it is.

Reply:
ah, sorry @jsirois didn't see your comment from earlier. sounds like the right approach is to remove the .truncate commit and go back to the explicit delete